### PR TITLE
Bug 800165 - Remove hidden elements from readability output

### PR DESCRIPTION
**Pull request by leibovic**
r? @gijsk @n1k0 

I still need to figure out how to solve the bustage this is causing with our jsdom tests. It looks like calling `node.style` sets an empty style attribute on the node, which is messing up our output. The thing is, we do have logic to remove style attributes in `_cleanStyles`, and I'm not sure why that's failing to remove these new attributes.

### CODE REVIEW

**Comment by n1k0** [comment]
This is indeed an issue with jsdom:

```
> doc = jsdom.jsdom("<html><body><p>plop</p></body></html>")
> p = doc.querySelector("p")
> p.style
> jsdom.serializeDocument(doc)
'<html><head></head><body><p style="">plop</p></body></html>'
```

Maybe cloning the element node before checking for its style would solve the issue? Though that's not supported by JSDOMParser atm, and will much likely impact performance. Will think of a better solution.

**Comment by n1k0** [comment]
Now I think about this feature, I'm not 100% sure we actually want to exclude whatever is hidden; think of an article of the following form:

```html
<article>
  <section>
    <p>Lorem…</p>
    <p><a href="#" onclick="$('#more-content').show()">Read more</a></p>
  </section>
  <section id="more-content" style="display:none">
    … ipsum dolor sit amet…
  </section>
```

File: Readability.js [hunk 1]
```diff
@@ -613,6 +613,12 @@ Readability.prototype = {
       var node = this._doc.documentElement;
 
       while (node) {
+        // Remove hidden elements.
+        if (node.style && node.style.display === "none") {
```
Comment (DavidBruant) [inline Readability.js]: Maybe also test the presence of the `hidden` attribute?
https://html.spec.whatwg.org/multipage/interaction.html#the-hidden-attribute

**Comment by gijsk** [comment]
> Now I think about this feature, I'm not 100% sure we actually want to exclude whatever is hidden; think of an article of the following form:

> ```html
<article>
  <section>
    <p>Lorem…</p>
    <p><a href="#" onclick="$('#more-content').show()">Read more</a></p>
  </section>
  <section id="more-content" style="display:none">
    … ipsum dolor sit amet…
  </section>
</article>
```

Yeah, this is tricky. I think we should err on the side of not including anything that was originally hidden. It's not possible for us to predict what content may become visible later, and clearly in this case the user can go to the website, click the link, and then click the reader button, and it'll Just Work. The reverse isn't true (ie if there's a bunch of hidden content, like the error messages on etherpad.m.o, for instance, then we don't do the right thing and the user can't fix that).

**Comment by gijsk** [comment]
Regarding the test issue, we should figure out why the style removal code isn't working.

**Comment by gijsk** [comment]
Huh, so changing @n1k0 's snippet to `p.removeAttribute('style')` does... nothing.

**Comment by n1k0** [comment]
This is definitely a bug in jsdom. The issue here is that the 3.x branch isn't maintained anymore, while it's the last one being supported with NodeJS (later versions require io.js).

We could definitely move to io.js, hoping we could then see this bug eventually addressed (latest 5.1.0 still has the bug), though that seems highly overkill :/

**Comment by n1k0** [comment]
Filed https://github.com/tmpvar/jsdom/issues/1109, let's see how it goes.

**Comment by gijsk** [comment]
jsdom has since been fixed and should now work with node rather than io.js. We should look at this again. :-)

**Comment by gijsk** [comment]
This got fixed in #283 / #285 . \o/

**Pull request closed without merging**